### PR TITLE
cargo-llvm-cov: fix tests, mark broken as needed

### PR DESCRIPTION
**Pull request by CobaltCause**
###### Description of changes

~~This packages [cargo-llvm-cov](https://github.com/taiki-e/cargo-llvm-cov), a tool for generating code coverage reports using built-in features of recent Rust toolchains.~~

~~Supersedes https://github.com/NixOS/nixpkgs/pull/197453~~
~~Alternative to https://github.com/NixOS/nixpkgs/pull/185183~~

Notably, this PR has all the tests running and passing, and does not wrap the binary, allowing users to choose their own toolchain versions.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by winterqt — CHANGES_REQUESTED** [review]
So the change I want to try here is using `fetchurl` and `postFetch` to not only get the lockfile from crates.io, but collapse it into a single FOD to take advantage of caching.

Essentially, use `fetchurl` on `https://crates.io/api/v1/crates/${pname}/${version}/download`, with `downloadToTemp` set to true, and do something like `tar xzf $downloadedFile ${pname}-${version}/Cargo.lock; mv */Cargo.lock $out` in `postFetch`:

```nix
fetchurl {
  name = "Cargo.lock";
  url = "https://crates.io/api/v1/crates/${pname}/${version}/download";
  downloadToTemp = true;
  postFetch = ''
    tar xzf $downloadedFile ${pname}-${version}/Cargo.lock
    mv */Cargo.lock $out
  '';
}
```

Then, copy this into the source in `postPatch` and `cargoUpdateHook`, which should then allow us to run tests while not having to include the lockfile.

**Comment by winterqt** [comment]
Though, now that I say that, I wonder if the compressed source tarball is smaller than the uncompressed Cargo.lock, lol. I'd guess that it's bigger, but who knows with the sourcery that is compression.

**Comment by CobaltCause** [comment]
Cursed: compress the checked-in `Cargo.lock` for maximum smallness.

I'll try implementing your suggestion. Is the concern mainly the size of the `nixpkgs` repository or the size of the stuff required to build the derivation?

File: pkgs/development/tools/rust/cargo-llvm-cov/default.nix
Comment (figsoda) [inline pkgs/development/tools/rust/cargo-llvm-cov/default.nix:22]: the crates.io version has a Cargo.lock, we can use `fetchCrate` to avoid the patch here

Comment (CobaltCause) [reply pkgs/development/tools/rust/cargo-llvm-cov/default.nix:22]: Yeah, but using `fetchCrate` doesn't pull in the fixtures needed to run the tests: https://github.com/NixOS/nixpkgs/pull/197453#issuecomment-1288327966

**Comment by winterqt** [comment]
I'd say checking in large autogenerated files is worse than doing this, but at the same time... this is a pretty small lockfile. Hrm.

Unrelated to this, but we currently [disable the profiler runtime on non-Linux platforms](https://github.com/NixOS/nixpkgs/blob/f157bef02b0902079561a02a2559e89ccb13236b/pkgs/development/compilers/rust/rustc.nix#L103). Given this, can you set `meta.broken = !(stdenv.isLinux && stdenv.targetPlatform.isRedox)`? (Maybe there's a better clause to use, not sure if we care about Redox here.)

(If you have Darwin hardware, feel free to try to enable it and test it -- if not, I'll do it sometime this week, I just don't want to have this package have to wait for a staging cycle.)

File: pkgs/development/tools/rust/cargo-llvm-cov/default.nix
Comment (winterqt) [reply pkgs/development/tools/rust/cargo-llvm-cov/default.nix:22]: This isn't possible, see https://github.com/NixOS/nixpkgs/pull/197453#issuecomment-1288327966 and https://github.com/NixOS/nixpkgs/pull/197453#issuecomment-1288334298.

**Comment by CobaltCause** [comment]
> I'd say checking in large autogenerated files is worse than doing this, but at the same time... this is a pretty small lockfile. Hrm.

Well, good news: your trick worked perfectly. And yeah I agree that checking in autogenerated stuff is suboptimal.

> Unrelated to this, but we currently [disable the profiler runtime on non-Linux platforms](https://github.com/NixOS/nixpkgs/blob/f157bef02b0902079561a02a2559e89ccb13236b/pkgs/development/compilers/rust/rustc.nix#L103). Given this, can you set meta.broken = !(stdenv.isLinux && stdenv.targetPlatform.isRedox)? (Maybe there's a better clause to use, not sure if we care about Redox here.)

Sure.

> (If you have Darwin hardware, feel free to try to enable it and test it -- if not, I'll do it sometime this week, I just don't want to have this package have to wait for a staging cycle.)

I don't, but I have a friends with x86_64-darwin and aarch64-darwin that I may be able to rope into this.

File: pkgs/development/tools/rust/cargo-llvm-cov/default.nix
Comment (winterqt) [inline pkgs/development/tools/rust/cargo-llvm-cov/default.nix:41]: This is an instance of [IFD](https://nixos.wiki/wiki/Import_From_Derivation), which is a no-go for Nixpkgs. Add the copy to `cargoUpdateHook`, and set `cargoVendorSha256` instead.

**Review by winterqt — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/rust/cargo-llvm-cov/default.nix
Comment (CobaltCause) [reply pkgs/development/tools/rust/cargo-llvm-cov/default.nix:41]: Honestly I'm kinda lost here. I understand how this is IFD and why that's not allowed (but didn't realize till you pointed it out), but searching for those things in my local copy Nixpkgs doesn't give me anything useful to go off of. I've tried a bunch of random things but can't get past this check: https://github.com/NixOS/nixpkgs/blob/f157bef02b0902079561a02a2559e89ccb13236b/pkgs/build-support/rust/fetch-cargo-tarball/default.nix#L44-L46

Comment (winterqt) [reply pkgs/development/tools/rust/cargo-llvm-cov/default.nix:41]: Right, because the Cargo update hook runs after that check, but before anything else, for whatever reason.

I personally don't see any downsides to moving it above that check -- do you mind making a commit to do that on this branch? Name it something like `fetchCargoTarball: run cargoUpdateHook before checking for Cargo.lock` or something, not sure. I'll ping someone re: that change after you push it just in case it's undesirable for whatever reason.

Comment (CobaltCause) [reply pkgs/development/tools/rust/cargo-llvm-cov/default.nix:41]: Hmm, that does fix it. It's weird to me that `postPatch = "cp ${cargoLock} Cargo.lock";` isn't enough to pass the check on its own though, and it looks like that's required even after this change too.

Comment (CobaltCause) [reply pkgs/development/tools/rust/cargo-llvm-cov/default.nix]: After this change, `nix-env -iA cargo-llvm-cov -f .` is spending a lot of time updating the crates.io index and downloading and vendoring crates, which the IFD version wasn't doing. I don't know if that means this change broke memoization completely or if it just busted an existing cache, but I thought it'd be worth mentioning.

Comment (winterqt) [reply pkgs/development/tools/rust/cargo-llvm-cov/default.nix]: > Hmm, that does fix it. It's weird to me that `postPatch = "cp ${cargoLock} Cargo.lock";` isn't enough to pass the check on its own though, and it looks like that's required even after this change too.

This is because the Cargo tarball derivation doesn't inherit the `postPatch` attribute, at least when using `buildRustPackage`.

> After this change, `nix-env -iA cargo-llvm-cov -f .` is spending a lot of time updating the crates.io index and downloading and vendoring crates, which the IFD version wasn't doing. I don't know if that means this change broke memoization completely or if it just busted an existing cache, but I thought it'd be worth mentioning.

This is expected -- the memoization of the crate files requires IFD. The `cargoLock.lockFile` mechanism and the vendoring mechanism are separate ways of retrieving the dependencies, one using IFD (and allowing for memoization) and one not.

**Comment by winterqt** [comment]
@Mic92 @zowoq Thoughts on 09b7abe1a29c4c47982984d8ea092b9c50031048, just to use `cargoUpdateHook` to copy a `Cargo.lock` into the derivation?

There's another option we can take here, and that's adding a way to specify a `postPatch` that gets passed to `fetchCargoTarball` when using `buildRustPackage`, but I think this is probably the least impactful (since only one other package even uses `cargoUpdateHook`) since we already have it (`cargoUpdateHook`).

File: pkgs/development/tools/rust/cargo-llvm-cov/default.nix
Comment (CobaltCause) [reply pkgs/development/tools/rust/cargo-llvm-cov/default.nix]: > This is because the Cargo tarball derivation doesn't inherit the `postPatch` attribute, at least when using `buildRustPackage`.

Yeah, I understand why it's needed, I was just noting that the UX was a little weird.

> This is expected -- the memoization of the crate files requires IFD. The `cargoLock.lockFile` mechanism and the vendoring mechanism are separate ways of retrieving the dependencies, one using IFD (and allowing for memoization) and one not.

Okay cool, just wanted to make sure I didn't screw anything up :sweat_smile:

Comment (winterqt) [reply pkgs/development/tools/rust/cargo-llvm-cov/default.nix]: > My understanding is that it has to be this way because of how vendoring works.

I honestly don't see what you mean -- as long as the patches result in the same lockfile being vendored as being added in the source, it shouldn't matter. Not sure why it was done like it is (e.g. separate `cargoPatches` array), sadly.

Comment (figsoda) [reply pkgs/development/tools/rust/cargo-llvm-cov/default.nix]: if i understood correctly, we can also use postFetch here inside the fetchFromGitHub

Comment (winterqt) [reply pkgs/development/tools/rust/cargo-llvm-cov/default.nix]: Yes, we can also do that.

However, I think we also shouldn't have this issue in the first place; ideally, the dependencies derivation would inherit the `postPatch` phase from the parent drv when using `buildRustPackage` -- or supply a separate option for it, like `cargoPatches`. But we have `cargoUpdateHook` so honestly I don't know. Maybe we should refactor how this works a bit? *shrug*

Comment (CobaltCause) [reply pkgs/development/tools/rust/cargo-llvm-cov/default.nix]: So I've just swapped over to @figsoda's `postFetch` suggestion, and this also obviates the need for the changes to `fetchCargoTarball`. I've left that commit in though because we probably want to do it anyway. Let me know if we want to drop it since it's not required anymore though.

> However, I think we also shouldn't have this issue in the first place; ideally, the dependencies derivation would inherit the `postPatch` phase from the parent drv when using `buildRustPackage` -- or supply a separate option for it, like `cargoPatches`. But we have `cargoUpdateHook` so honestly I don't know. Maybe we should refactor how this works a bit? _shrug_

I am in agreement with all of this, lol

**Comment by CobaltCause** [comment]
Looks like someone else beat us to it: https://github.com/NixOS/nixpkgs/pull/192813

Do we want to close this, or should I rebase on top of their changes so we can have working tests and a proper `meta.broken`?

**Comment by winterqt** [comment]
Let's do that, and adjust the commit message(s?) appropriately, please.

Sorry about forgetting this.

**Comment by CobaltCause** [comment]
No problem! I can do that, yes. Do we want to leave the changes to `fetchCargoTarball` or get rid of it since it's not needed anymore?

**Comment by winterqt** [comment]
> Do we want to leave the changes to fetchCargoTarball or get rid of it since it's not needed anymore?

Probably best to move it to a separate PR, so this isn't blocked on any (potential) issues/discussions.

**Comment by CobaltCause** [comment]
Okay, rebase is done and `fetchCargoTarball` changes are gone. The diff is pretty big because I just kind of blindly rebased, should I break it into smaller commits?

**Comment by winterqt** [comment]
Looks good, except for one thing I think is worth mentioning: what do you think about splitting this into two (three...?) commits: one that fixes the tests, one that marks it as broken, and *maybe* another that adds you as a maintainer (unless you want to roll that one into the first one)?

Not sure, is that too nitpicky?

**Comment by CobaltCause** [comment]
Not too nitpicky, no; I think that's fair. I would have done it that way originally but I wasn't sure if having a bunch of commits was desirable. I've split it into three commits.

**Comment by wucke13** [comment]
@GrahamcOfBorg build cargo-llvm-cov

**Comment by winterqt** [comment]
@wucke13 FWIW, OfBorg automatically builds packages when commit messages are formatted properly.

Looks like tests are failing now because of invalid `llvm-profdata` arguments. 😕

Can look into this when I have the time, or anyone else can.

**Comment by winterqt** [comment]
I... cannot reproduce this locally! What even?

@ofborg eval

**Comment by CobaltCause** [comment]
Rebased this, but now the tests are failing. I don't really understand why, maybe something to do with temporary directories?

**Comment by winterqt** [comment]
Forgot to exclude a file from your tree? :)

What was the issue? Sorry, haven't had time to look into this.

**Comment by CobaltCause** [comment]
> Forgot to exclude a file from your tree? :)

Oh lmao, a careless `git add` of something for a different PR. Thanks.

> What was the issue? Sorry, haven't had time to look into this.

It's alright. After the rebase to get the updates other people contributed, the tests are failing. I believe this is the relevant output:

```
error: no input files specified. See llvm-profdata merge -help
error: failed to merge profile data: process didn't exit successfully: `/nix/store/36wl38al1d4k0762d5d78xj3h8wl3ygw-llvm-14.0.6/bin/llvm-profdata merge -sparse -f /build/.tmpWy15FV -o /build/.tmpfw0DD9/target/llvm-cov-target/.tmpfw0DD9.profdata` (exit status: 1)
```

The string "failed to merge profile data" comes from `cargo-llvm-cov` itself, I think the lack of input files has to do with the `/build/.tmpWy15FV` directory specifically. Not exactly sure how to get a look in there while testing in the sandbox, though. What's weird is that `cargo test` after setting up a shell in the project itself works fine.

**Comment by winterqt** [comment]
I tried reverting this back to 0.5.0 (which is the version this PR originally had the tests working on), and it throws the same error... 😕

Not really sure what could be up here. It seems like `llvm-profdata` can also throw that when the files exist, but aren't valid?

Upstream's CI works fine.

@taiki-e Do you happen to have any idea what could be up here, by chance? Thank you!

**Comment by winterqt** [comment]
Also, @CobaltCause, you reverted the hash updates in your rebase -- please fix them. Thanks!

**Comment by taiki-e** [comment]
> @taiki-e Do you happen to have any idea what could be up here, by chance? Thank you!

What version of Cargo/rustc/LLVM is used in the test? Specifying an incompatible version of LLVM in LLVM_COV/LLVM_PROPDATA may result in failure to generate coverage.

(Older LLVM may work, but it is recommended to use the LLVM version that is the same as the one displayed by `rustc -vV`.  Neither rustc nor cargo-llvm-cov has done much testing for compatibility with older LLVM...)

**Comment by winterqt** [comment]
> What version of Cargo/rustc/LLVM is used in the test? Specifying an incompatible version of LLVM in LLVM_COV/LLVM_PROPDATA may result in failure to generate coverage.

rustc/Cargo 1.66.1, and LLVM 14.0.6 (which is both the version that we get llvm-cov and llvm-profdata from, as well as the version we compile rustc with).

I wonder why an error isn't being thrown by rustc if the coverage couldn't be generated, if that's the case...

**Comment by CobaltCause** [comment]
> Also, @CobaltCause, you reverted the hash updates in your rebase -- please fix them. Thanks!

Whoops! Fixed.

**Comment by CobaltCause** [comment]
I rebased this yesterday and now all the checks are green :thinking:

**Comment by winterqt** [comment]
Wanna bisect to see what fixed it? I suspect one of the recent Cargo configuration changes, but I'm curious :)

**Comment by CobaltCause** [comment]
Yes, I think that's a good idea. I'd've done it sooner but I had some urgent IRL stuff going on. Anyway, I just finished bisecting now.

<details>
<summary><code>git bisect log</code></summary>

```
git bisect start '--term-new=fixed' '--term-old=unfixed'
# status: waiting for both good and bad commits
# fixed: [a0527721cf656e1eaef11f51ab672ea3268ce3c7] Merge pull request #230902 from Aleksanaa/hcledit
git bisect fixed a0527721cf656e1eaef11f51ab672ea3268ce3c7
# status: waiting for good commit(s), bad commit known
# unfixed: [edee331bc786997a0d47fff9fccdcb3895afae69] Merge pull request #217522 from r-ryantm/auto-update/rtsp-simple-server
git bisect unfixed edee331bc786997a0d47fff9fccdcb3895afae69
# fixed: [b5f452b688cf54dbe12588563838387a03b8a4a1] Merge staging-next into staging
git bisect fixed b5f452b688cf54dbe12588563838387a03b8a4a1
# unfixed: [daed54a9d4c287fbf0940510927d23913922fe2e] Merge pull request #221138 from fabaff/sshfs
git bisect unfixed daed54a9d4c287fbf0940510927d23913922fe2e
# unfixed: [dc400218aa3aebb0908c4a7b1b7e25155242cf00] Merge pull request #222446 from figsoda/nix-init
git bisect unfixed dc400218aa3aebb0908c4a7b1b7e25155242cf00
# unfixed: [8c4c762da229b5f7c83e3946eab8875df9dd88c6] lighttpd: Disable tests for DES and MD5
git bisect unfixed 8c4c762da229b5f7c83e3946eab8875df9dd88c6
# unfixed: [dcd028c9e01b51caee8cda37b231c36b64af7d42] Merge pull request #222998 from viraptor/livepeer-darwin
git bisect unfixed dcd028c9e01b51caee8cda37b231c36b64af7d42
# unfixed: [5d9c49eedc3227f9cca32a3ea5ec351740ef0960] Merge pull request #223884 from fabaff/peaqevcore-bump
git bisect unfixed 5d9c49eedc3227f9cca32a3ea5ec351740ef0960
# unfixed: [0553481d359d129f8839fa870b516aba2a6e0fd6] Merge pull request #224031 from r-ryantm/auto-update/python310Packages.timezonefinder
git bisect unfixed 0553481d359d129f8839fa870b516aba2a6e0fd6
# fixed: [80563f40314a890ee0c9a4957a663dd2bc1f7585] Merge staging-next into staging
git bisect fixed 80563f40314a890ee0c9a4957a663dd2bc1f7585
# fixed: [c81880f0a497d99661895a99cb413f5f783843f9] Merge #221422: dbus: 1.14.4 -> 1.14.6
git bisect fixed c81880f0a497d99661895a99cb413f5f783843f9
# fixed: [1d9429faa0c864cae4e4f366818d47b18c32234e] polkit: move test-only deps to nativeCheckInputs
git bisect fixed 1d9429faa0c864cae4e4f366818d47b18c32234e
# unfixed: [a0537d633fdbd90f26808df5ca0a890cb4d6ebdf] apple_sdk_11_0: simplify SDK unpacking
git bisect unfixed a0537d633fdbd90f26808df5ca0a890cb4d6ebdf
# fixed: [00623e191a042012068a29f93cb1d4ac635b5a38] Merge staging-next into staging
git bisect fixed 00623e191a042012068a29f93cb1d4ac635b5a38
# fixed: [d681bc866130f10211cb8ba33d4826de5f023f3d] Merge staging-next into staging
git bisect fixed d681bc866130f10211cb8ba33d4826de5f023f3d
# fixed: [7d23bdd221d219087f07e83b8486d0e93e625543] Merge pull request #221413 from alyssais/rustc-musl
git bisect fixed 7d23bdd221d219087f07e83b8486d0e93e625543
# fixed: [470e6130b347c5aabc63cd6600ba78fa909e8e2a] rust: fix overriding rust flags on musl
git bisect fixed 470e6130b347c5aabc63cd6600ba78fa909e8e2a
# first fixed commit: [470e6130b347c5aabc63cd6600ba78fa909e8e2a] rust: fix overriding rust flags on musl
```

</details>

TL;DR: 470e6130b347c5aabc63cd6600ba78fa909e8e2a is the first fixed commit

Looks like I didn't screw up my bisection this time, lol

File: pkgs/development/tools/rust/cargo-llvm-cov/default.nix
Comment (figsoda) [inline pkgs/development/tools/rust/cargo-llvm-cov/default.nix:43]: might be better to put this in `postFetch` of the outer derivation

Comment (figsoda) [reply pkgs/development/tools/rust/cargo-llvm-cov/default.nix:43]: this wasn't possible when the PR was created, might it should be good now thanks to #221093

File: pkgs/development/tools/rust/cargo-llvm-cov/default.nix
Comment (figsoda) [inline pkgs/development/tools/rust/cargo-llvm-cov/default.nix:44]:
```suggestion
    sha256 = "sha256-+ddGNS2NcYNslkgd6DmJr0gtEvF49oOWTeDoNMJ8yIc=";
```

File: pkgs/development/tools/rust/cargo-llvm-cov/default.nix
Comment (figsoda) [inline pkgs/development/tools/rust/cargo-llvm-cov/default.nix:48]:
```suggestion
  cargoSha256 = "sha256-eYNb3FJkOxwyWAs8fEyQOEqpD3Fw5tCokdh5ghO8kro=";

  # Upstream doesn't include the lockfile so we need to add it back
  postUnpack = ''
    cp ${cargoLock} source/Cargo.lock
  '';
```

File: pkgs/development/tools/rust/cargo-llvm-cov/default.nix
Comment (CobaltCause) [reply pkgs/development/tools/rust/cargo-llvm-cov/default.nix]: Thanks for spelling it out for me in https://github.com/NixOS/nixpkgs/pull/197478#pullrequestreview-1574403652 since I wasn't really sure what you were asking for lol. I've just applied this suggestion.

**Comment by figsoda** [comment]
oops, merge conflict again

File: pkgs/development/tools/rust/cargo-llvm-cov/default.nix
Comment (winterqt) [inline pkgs/development/tools/rust/cargo-llvm-cov/default.nix:51]: I don't think it should go here, but can you add a similar note about why the tests would break in the future (referencing the conversations we've had here + the upstream issue I filed)?

**Review by winterqt — DISMISSED** [review]

File: pkgs/development/tools/rust/cargo-llvm-cov/default.nix
Comment (CobaltCause) [reply pkgs/development/tools/rust/cargo-llvm-cov/default.nix:68]: Done in a new commit, lmk whether what I wrote is adequate

**Review by matthiasbeyer — APPROVED** [review]

**Review by matthiasbeyer — APPROVED** [review]
I think we can merge this now, can't we?

**Comment by CobaltCause** [comment]
I mean, I think so, but I'm obviously biased. And it would be nice to get an approval from @winterqt since they've invested a lot of time helping with this and reviewing it though.

**Review by matthiasbeyer — APPROVED** [review]

**Comment by matthiasbeyer** [comment]
Result of `nixpkgs-review pr 197478` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>cargo-llvm-cov</li>
  </ul>
</details>

**Pull request merged**
